File: data-provider/src/main/scala/velocorner/model/DailyProgress.scala
package velocorner.model

import org.joda.time.LocalDate
import velocorner.api.{Activity, Progress}


object DailyProgress {

  def fromStorage(activity: Activity): DailyProgress = {
    val progress = new Progress(1,
      activity.distance / 1000, activity.distance / 1000, activity.moving_time,
      activity.average_speed.getOrElse(0f).toDouble,
      activity.total_elevation_gain, activity.total_elevation_gain
    )
    DailyProgress(activity.start_date_local.toLocalDate, progress)
  }

  def fromStorage(activities: Iterable[Activity]): Iterable[DailyProgress] = {
    activities.map(fromStorage)
      .groupBy(_.day)
      .map{ case (day, progressPerDay) =>
        DailyProgress(day, progressPerDay.foldLeft(Progress.zero)((accu, dailyProgress) => accu + dailyProgress.progress))
      }.toSeq.sortBy(_.day.toString)
  }

  def aggregate(list: Iterable[DailyProgress]): Iterable[DailyProgress] = {
    list.scanLeft(DailyProgress(LocalDate.now, Progress.zero))((accu, i) =>
      DailyProgress(i.day, accu.progress + i.progress)).tail
  }
}


case class DailyProgress(day: LocalDate, progress: Progress)

File: web-backend/src/main/kotlin/velocorner/backend/model/DailyProgress.kt
package velocorner.backend.model

import java.time.LocalDate

data class DailyProgress(val day: LocalDate, val progress: Progress) {

    companion object {
        fun from(activity: Activity): DailyProgress {
            val progress = Progress(
                days = 1,
                rides = 1,
                distance = activity.distance / 1000,
                longestDistance = activity.distance / 1000,
                movingTime = activity.moving_time.toLong(),
                averageSpeed = activity.average_speed?.toDouble() ?: 0.0,
                elevation = activity.total_elevation_gain,
                longestElevation = activity.total_elevation_gain
            )
            return DailyProgress(activity.getStartDateLocal().toLocalDate(), progress)
        }

        fun from(activities: Sequence<Activity>): List<DailyProgress> =
            activities
                .map { from(it) }
                .groupBy { it.day }
                .map { (day, progressPerDay) ->
                    DailyProgress(
                        day = day,
                        progress = progressPerDay
                            .fold(Progress.zero) { accu, dailyProgress ->
                                accu + dailyProgress.progress
                            }
                            .copy(days = 1)
                    )
                }
                .sortedBy { it.day.toString() }

        fun aggregate(list: Iterable<DailyProgress>): List<DailyProgress> =
            list.scan(DailyProgress(LocalDate.now(), Progress.zero)) { accu, item ->
                DailyProgress(item.day, accu.progress + item.progress)
            }.drop(1)
    }
}

